// handsontable/validators/index.mjs
import { autocompleteValidator, VALIDATOR_TYPE as AUTOCOMPLETE_VALIDATOR } from "./autocompleteValidator/index.mjs";
import { dateValidator, VALIDATOR_TYPE as DATE_VALIDATOR } from "./dateValidator/index.mjs";
import { dropdownValidator, VALIDATOR_TYPE as DROPDOWN_VALIDATOR } from "./dropdownValidator/index.mjs";
import { intlDateValidator, VALIDATOR_TYPE as INTL_DATE_VALIDATOR } from "./intlDateValidator/index.mjs";
import { intlTimeValidator, VALIDATOR_TYPE as INTL_TIME_VALIDATOR } from "./intlTimeValidator/index.mjs";
import { multiSelectValidator, VALIDATOR_TYPE as MULTISELECT_VALIDATOR } from "./multiSelectValidator/index.mjs";
import { numericValidator, VALIDATOR_TYPE as NUMERIC_VALIDATOR } from "./numericValidator/index.mjs";
import { timeValidator, VALIDATOR_TYPE as TIME_VALIDATOR } from "./timeValidator/index.mjs";
import { registerValidator } from "./registry.mjs";
/**
 * Registers all available validators.
 */
export function registerAllValidators() {
  registerValidator(autocompleteValidator);
  registerValidator(dropdownValidator);
  registerValidator(dateValidator);
  registerValidator(intlDateValidator);
  registerValidator(intlTimeValidator);
  registerValidator(multiSelectValidator);
  registerValidator(numericValidator);
  registerValidator(timeValidator);
}
export { autocompleteValidator, AUTOCOMPLETE_VALIDATOR, dropdownValidator, DROPDOWN_VALIDATOR, dateValidator, DATE_VALIDATOR, intlDateValidator, INTL_DATE_VALIDATOR, intlTimeValidator, INTL_TIME_VALIDATOR, multiSelectValidator, MULTISELECT_VALIDATOR, numericValidator, NUMERIC_VALIDATOR, timeValidator, TIME_VALIDATOR };
export { getRegisteredValidatorNames, getRegisteredValidators, getValidator, hasValidator, registerValidator } from "./registry.mjs";
