// handsontable/renderers/intlDateRenderer/intlDateRenderer.js
"use strict";

exports.__esModule = true;
exports.intlDateRenderer = intlDateRenderer;
exports.valueFormatter = valueFormatter;
var _textRenderer = require("../textRenderer");
var _dateTime = require("../../helpers/dateTime");
var _mixed = require("../../helpers/mixed");
var _object = require("../../helpers/object");
var _constants = require("../../helpers/constants");
const RENDERER_TYPE = exports.RENDERER_TYPE = 'intl-date';
const DEFAULT_INTL_FORMAT = {
  year: 'numeric',
  month: '2-digit',
  day: '2-digit'
};

/**
 * Formats the value using the date format.
 *
 * @param {*} value Value to be formatted.
 * @param {CellMeta} cellProperties Cell meta object.
 * @returns {*} Returns the rendered value.
 */
function valueFormatter(value, cellProperties) {
  const {
    dateFormat,
    locale,
    allowEmpty
  } = cellProperties;
  if ((0, _mixed.isEmpty)(value)) {
    return allowEmpty ? value : _constants.BAD_VALUE_TEXT;
  }
  const date = (0, _dateTime.parseToLocalDate)(value);
  if (date === null) {
    return _constants.BAD_VALUE_TEXT;
  }
  const intlFormat = (0, _object.isObject)(dateFormat) ? dateFormat : DEFAULT_INTL_FORMAT;
  return new Intl.DateTimeFormat(locale, intlFormat).format(date);
}

/**
 * Handsontable renderer.
 *
 * @private
 * @param {Core} hotInstance The Handsontable instance.
 * @param {HTMLTableCellElement} TD The rendered cell element.
 * @param {number} row The visual row index.
 * @param {number} col The visual column index.
 * @param {number|string} prop The column property (passed when datasource is an array of objects).
 * @param {*} value The rendered value.
 * @param {object} cellProperties The cell meta object (see {@link Core#getCellMeta}).
 */
function intlDateRenderer(hotInstance, TD, row, col, prop, value, cellProperties) {
  _textRenderer.textRenderer.apply(this, [hotInstance, TD, row, col, prop, value, cellProperties]);
}
intlDateRenderer.valueFormatter = valueFormatter;
intlDateRenderer.RENDERER_TYPE = RENDERER_TYPE;
